File: javascript/sdk/webapi.js
/**
 * WebApi module - Subscription activation via web API
 * Equivalent to PHP WebApi class in sdk_file.php
 */

const Core = require('./core');
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

class WebApi extends Core {
  /**
   * @param {string} applicationId - Application ID (or read from appid.txt)
   * @param {string} serviceProviderAppId - Service provider app ID
   * @param {string} hashSecret - Secret for SHA256 signature
   */
  constructor(applicationId = null, serviceProviderAppId = 'DSAPP_000003', hashSecret = '8c899b6b56e6855605ea61be994012e7') {
    super();
    this.url = 'https://developer.bdapps.com/subscription/activate';
    this.requestId = String(Math.floor(Math.random() * 900000000000000000) + 100000000000000000);
    this.requestTime = this._getTime();
    this.serviceProviderAppId = serviceProviderAppId;
    this.hashSecret = hashSecret;
    this.applicationId = applicationId;
    this.subscriberId = null;
    this.requestSignature = null;
  }

  /** @private */
  _getTime() {
    // Asia/Dhaka timezone format: YYYY-MM-DD HH:mm:ss.000
    const now = new Date();
    const options = { timeZone: 'Asia/Dhaka' };
    const year = now.toLocaleString('en-US', { ...options, year: 'numeric' });
    const month = now.toLocaleString('en-US', { ...options, month: '2-digit' });
    const day = now.toLocaleString('en-US', { ...options, day: '2-digit' });
    const hour = now.toLocaleString('en-US', { ...options, hour: '2-digit', hour12: false });
    const minute = now.toLocaleString('en-US', { ...options, minute: '2-digit' });
    const second = now.toLocaleString('en-US', { ...options, second: '2-digit' });
    return `${year}-${month}-${day} ${hour}:${minute}:${second}.000`;
  }

  /**
   * Set the subscriber and compute the request signature
   * @param {string} address - Subscriber address
   */
  getAppAndSubscriber(address) {
    this.subscriberId = address;

    // Try to read applicationId from appid.txt if not set
    if (!this.applicationId) {
      try {
        const appIdFile = path.resolve('appid.txt');
        this.applicationId = fs.readFileSync(appIdFile, 'utf8').trim();
      } catch {
        // applicationId must be set via constructor or env
      }
    }

    // Generate SHA256 signature
    this.requestSignature = crypto
      .createHash('sha256')
      .update(this.applicationId + this.requestTime + this.hashSecret)
      .digest('hex');
  }

  /**
   * Send the subscription activation request
   * @returns {Promise<object>} Response from BDApps
   */
  async requestSend() {
    const payload = {
      requestId: this.requestId,
      requestTime: this.requestTime,
      serviceProviderAppId: this.serviceProviderAppId,
      requestSignature: this.requestSignature,
      applicationId: this.applicationId,
      subscriberId: this.subscriberId,
    };

    const response = await this.sendRequest(payload, this.url);
    return response;
  }
}

module.exports = WebApi;

File: python/sdk/webapi.py
"""
WebApi module - Subscription activation via web API.
Equivalent to PHP WebApi class in sdk_file.php.
"""

import hashlib
import os
import random
from datetime import datetime
from typing import Optional

from .core import Core


class WebApi(Core):
    """
    Subscription activation via web API.
    Equivalent to PHP WebApi class.
    """

    def __init__(
        self,
        application_id: Optional[str] = None,
        service_provider_app_id: str = "DSAPP_000003",
        hash_secret: str = "8c899b6b56e6855605ea61be994012e7",
    ):
        super().__init__()
        self.url = "https://developer.bdapps.com/subscription/activate"
        self.request_id = str(random.randint(100000000000000000, 999999999999999999))
        self.request_time = self._get_time()
        self.service_provider_app_id = service_provider_app_id
        self.hash_secret = hash_secret
        self.application_id = application_id
        self.subscriber_id: Optional[str] = None
        self.request_signature: Optional[str] = None

    def _get_time(self) -> str:
        """Get current time in Asia/Dhaka timezone format."""
        # Format: YYYY-MM-DD HH:MM:SS.000
        now = datetime.now()
        return now.strftime("%Y-%m-%d %H:%M:%S.000")

    def get_app_and_subscriber(self, address: str) -> None:
        """
        Set the subscriber and compute the request signature.

        Args:
            address: Subscriber address.
        """
        self.subscriber_id = address

        # Try to read applicationId from appid.txt if not set
        if not self.application_id:
            try:
                app_id_file = os.path.abspath("appid.txt")
                with open(app_id_file, "r") as f:
                    self.application_id = f.readline().strip()
            except FileNotFoundError:
                pass  # applicationId must be set via constructor or env

        # Generate SHA256 signature
        sig_input = f"{self.application_id}{self.request_time}{self.hash_secret}"
        self.request_signature = hashlib.sha256(sig_input.encode()).hexdigest()

    async def request_send(self) -> dict:
        """
        Send the subscription activation request.

        Returns:
            Response from BDApps.
        """
        payload = {
            "requestId": self.request_id,
            "requestTime": self.request_time,
            "serviceProviderAppId": self.service_provider_app_id,
            "requestSignature": self.request_signature,
            "applicationId": self.application_id,
            "subscriberId": self.subscriber_id,
        }

        return await self.send_request(payload, self.url)
